src/context/TodoContext.tsx

```diff
@@ -0,0 +1,141 @@
+import React, {
+  Dispatch,
+  SetStateAction,
+  createContext,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import { Todo } from '../types/Todo';
+import { ErrorList } from '../types/ErrorList';
+import { Status } from '../types/Status';
+import { wait } from '../utils/fetchClient';
+import { deleteTodo, getTodos } from '../api/todos';
+
+export type TodoContext = {
+  todos: Todo[];
+  preparedTodos: Todo[];
+  setTodos: React.Dispatch<React.SetStateAction<Todo[]>>;
+  tempTodo: Todo | null;
+  setTempTodo: Dispatch<SetStateAction<Todo | null>>;
+  errorMessage: ErrorList | string;
+  setErrorMessage: Dispatch<SetStateAction<ErrorList | string>>;
+  status: Status;
+  setStatus: (value: Status) => void;
+  handleError: (value: string) => void;
+  query: string;
+  setQuery: (value: string) => void;
+
+  isLoading: boolean;
+  setIsLoading: (value: boolean) => void;
+  focusInput: () => void;
+  handleDeleteTodo: (ItemId: number) => void;
+  processingIds: number[];
+  setProcessingIds: (value: number[]) => void;
+  contextInputRef: React.MutableRefObject<HTMLInputElement | null>;
+};
+
+export const TodosContext = createContext<TodoContext | undefined>(undefined);
+
+export const useTodosContext = () => {
+  const context = useContext(TodosContext);
+
+  if (context === undefined) {
+    throw new Error('Error');
+  }
+
+  return context;
+};
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState<ErrorList | string>('');
+  const [status, setStatus] = useState<Status>(Status.All);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [query, setQuery] = useState<string>('');
+
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [processingIds, setProcessingIds] = useState<number[]>([]);
+  const contextInputRef = useRef<HTMLInputElement | null>(null);
```

Comment:
I'm not sure that's a good place for input ref. 

It's not related to todos logic and is more about some specific UI, so let's move it closer to it